x-pack/plugins/infra/public/pages/metrics/hosts/hooks/use_control_panels_url_state.ts

```diff
@@ -6,48 +6,117 @@
+      fieldName: availableControlsPanels.HOST_OS_NAME,
+      title: 'Operating System',
     },
-    cloudProviderPanel: {
-      order: 1,
-      width: 'medium',
-      grow: false,
-      type: 'optionsListControl',
-      explicitInput: {
-        id: 'cloudProviderPanel',
-        dataViewId,
-        fieldName: 'cloud.provider',
-        title: 'Cloud Provider',
-      },
+  },
+  [availableControlsPanels.CLOUD_PROVIDER]: {
+    order: 1,
+    width: 'medium',
+    grow: false,
+    type: 'optionsListControl',
+    explicitInput: {
+      id: availableControlsPanels.CLOUD_PROVIDER,
+      fieldName: availableControlsPanels.CLOUD_PROVIDER,
+      title: 'Cloud Provider',
     },
-  } as unknown as ControlGroupInput['panels']);
-const HOST_FILTERS_URL_STATE_KEY = 'controlPanels';
+  },
+};
+
+const availableControlPanelFields = Object.values(availableControlsPanels);
 
-export const useControlPanels = (dataViewId: string) => {
-  return useUrlState<ControlPanels>({
-    defaultState: getDefaultPanels(dataViewId),
+export const useControlPanels = (
+  dataView: DataView
+): [ControlPanels, (state: ControlPanels) => void] => {
+  const defaultState = getVisibleControlPanelsConfig(dataView);
+
+  const [controlPanels, setControlPanels] = useUrlState<ControlPanels>({
+    defaultState,
     decodeUrlState,
     encodeUrlState,
     urlStateKey: HOST_FILTERS_URL_STATE_KEY,
   });
+
+  /**
+   * Configure the control panels as
+   * 1. Available fields from the data view
+   * 2. Existing filters from the URL parameter (not colliding with allowed fields from data view)
+   * 3. Enhanced with dataView.id
+   */
+  const controlsPanelsWithId = mergeDefaultPanelsWithUrlConfig(dataView, controlPanels);
+
+  return [controlsPanelsWithId, setControlPanels];
+};
+
+/**
+ * Utils
+ */
+
+const getVisibleControlPanels = (dataView: DataView) => {
+  return availableControlPanelFields.filter(
+    (panelKey) => dataView.fields.getByName(panelKey) !== undefined
+  );
+};
+
+const getVisibleControlPanelsConfig = (dataView: DataView) => {
+  return getVisibleControlPanels(dataView).reduce((panelsMap, panelKey) => {
+    const config = controlPanelConfigs[panelKey];
+
+    return { ...panelsMap, [panelKey]: config };
+  }, {});
+};
+
+const addDataViewIdToControlPanels = (controlPanels: ControlPanels, dataViewId: string = '') => {
+  return _.mapValues(controlPanels, (controlPanelConfig) => {
+    const controlsClone = _.cloneDeep(controlPanelConfig);
+    controlsClone.explicitInput.dataViewId = dataViewId;
+    return controlsClone;
+  });
+};
+
+const cleanControlPanels = (controlPanels: ControlPanels) => {
+  return _.mapValues(controlPanels, (controlPanelConfig) => {
+    const controlsClone = _.cloneDeep(controlPanelConfig);
+    delete controlsClone.explicitInput.dataViewId;
+    return controlsClone;
+  });
+};
+
+const mergeDefaultPanelsWithUrlConfig = (dataView: DataView, urlPanels: ControlPanels = {}) => {
```

Comment:
nit: Maybe the methods can be defined before the usage in this file - I think the readability will be also better that way